File: tools-manage/invocationlab-admin/src/main/java/org/javamaster/invocationlab/admin/filter/CompressFilter.java
package org.javamaster.invocationlab.admin.filter;

import com.google.common.collect.Sets;
import lombok.Cleanup;
import org.springframework.http.HttpHeaders;

import javax.servlet.Filter;
import javax.servlet.FilterChain;
import javax.servlet.FilterConfig;
import javax.servlet.ServletException;
import javax.servlet.ServletOutputStream;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.annotation.WebFilter;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Set;
import java.util.zip.GZIPOutputStream;

/**
 * @author yudong
 * @date 2023/8/5
 */
@WebFilter(urlPatterns = "/*")
public class CompressFilter implements Filter {
    private Set<String> excludePaths;

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain) throws IOException, ServletException {
        HttpServletRequest req = (HttpServletRequest) request;
        HttpServletResponse res = (HttpServletResponse) response;

        res.addHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_CREDENTIALS, "false");
        res.addHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS, "*");
        res.addHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_METHODS, "*");
        res.addHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*");
        res.addHeader(HttpHeaders.ACCESS_CONTROL_EXPOSE_HEADERS, HttpHeaders.CONTENT_DISPOSITION);

        String requestURI = req.getRequestURI();
        if (excludePaths.contains(requestURI)) {
            filterChain.doFilter(req, res);
            return;
        }
        if (!isGipSupported(req)) {
            filterChain.doFilter(req, res);
            return;
        }
        res.setHeader("Content-Encoding", "gzip");
        ResponseWrapper responseWrapper = new ResponseWrapper(res);
        filterChain.doFilter(req, responseWrapper);
        @Cleanup
        ServletByteArrayOutputStream originalOutputStream = (ServletByteArrayOutputStream) responseWrapper.getOutputStream();
        @Cleanup
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        @Cleanup
        GZIPOutputStream gzipOutputStream = new GZIPOutputStream(byteArrayOutputStream);
        gzipOutputStream.write(originalOutputStream.toByteArray());
        gzipOutputStream.finish();
        ServletOutputStream realOutPutStream = res.getOutputStream();
        byteArrayOutputStream.writeTo(realOutPutStream);
        realOutPutStream.flush();
    }


    private boolean isGipSupported(HttpServletRequest req) {
        String encoding = req.getHeader("Accept-Encoding");
        return encoding != null && encoding.contains("gzip");
    }

    @Override
    public void init(FilterConfig filterConfig) {
        excludePaths = Sets.newHashSet(
                "/ncnb/project/exportErd",
                "/ncnb/queryInfo/exportSql",
                "/actuator/health"
        );
    }

}

File: tools-manage/invocationlab-admin-21/src/main/kotlin/org/javamaster/invocationlab/admin/filter/CompressFilter.kt
package org.javamaster.invocationlab.admin.filter

import com.google.common.collect.Sets
import jakarta.servlet.*
import jakarta.servlet.annotation.WebFilter
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpHeaders
import java.io.ByteArrayOutputStream
import java.util.zip.GZIPOutputStream

/**
 * @author yudong
 * @date 2023/8/5
 */
@WebFilter(urlPatterns = ["/*"])
class CompressFilter : Filter {
    private lateinit var excludePaths: Set<String>

    override fun doFilter(request: ServletRequest, response: ServletResponse, filterChain: FilterChain) {
        val req = request as HttpServletRequest
        val res = response as HttpServletResponse

        res.addHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_CREDENTIALS, "false")
        res.addHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS, "*")
        res.addHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_METHODS, "*")
        res.addHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*")
        res.addHeader(HttpHeaders.ACCESS_CONTROL_EXPOSE_HEADERS, HttpHeaders.CONTENT_DISPOSITION)

        val requestURI = req.requestURI
        if (excludePaths.contains(requestURI)) {
            filterChain.doFilter(req, res)
            return
        }
        if (!isGipSupported(req)) {
            filterChain.doFilter(req, res)
            return
        }
        res.setHeader("Content-Encoding", "gzip")
        val responseWrapper = ResponseWrapper(res)
        filterChain.doFilter(req, responseWrapper)
        (responseWrapper.outputStream as ServletByteArrayOutputStream).use { originalOutputStream ->
            val byteArrayOutputStream = ByteArrayOutputStream()
            GZIPOutputStream(byteArrayOutputStream).use {
                it.write(originalOutputStream.toByteArray())
                it.finish()
                val realOutPutStream = res.outputStream
                byteArrayOutputStream.writeTo(realOutPutStream)
                realOutPutStream.flush()
            }
        }
    }


    private fun isGipSupported(req: HttpServletRequest): Boolean {
        val encoding = req.getHeader("Accept-Encoding")
        return encoding != null && encoding.contains("gzip")
    }

    override fun init(filterConfig: FilterConfig) {
        excludePaths = Sets.newHashSet(
            "/ncnb/project/exportErd",
            "/ncnb/queryInfo/exportSql",
            "/actuator/health"
        )
    }
}
